### PR TITLE
Add AtomInfo.get_building_block_atom()

### PR DESCRIPTION
**Pull request by lukasturcani**
Should address https://github.com/lukasturcani/stk/issues/260

Main things:

* Add `AtomInfo.get_building_block_atom()`.
* Add `Molecule.get_canonical_atom_ids()`.

Adding `Molecule.get_canonical_atom_ids()` is a necessary evil, as far as I can tell. Here is why:

* `ConstructedMolecule.with_canonical_atom_ordering()` must work.
* When `ConstructedMolecule.with_canonical_atom_ordering()` is called, it also calls `Molecule.with_canonical_atom_ordering()` on its building blocks. This means the building blocks of a canonical constructed molecule are replaced with canonical building blocks.
* After `ConstructedMolecule.with_canonical_atom_ordering()` is called, `AtomInfo.get_building_block()` points to the new, canonical building block.
* Similarly `AtomInfo.get_building_block_atom()` needs to point to the atom in the new, canonical building block.
* The implementation of `ConstructedMolecule.with_canonical_atom_ordering()` which creates new `AtomInfo` instances for the new, canonical constructed molecule, needs to take the `Atom` from the old `AtomInfo` and figure out what the corresponding `Atom` in the new, canonical building block is. For this, the building block needs to provide a map, which maps the old id to the canonical id.

I will note that there is a slight inefficiency which I'm not too happy about. Canonical ordering is computed twice for each building block. Once when `with_canonical_atom_ordering()` is  called on it, and once when `get_canonical_atom_ids()` is called. This can be addressed in two ways as far as I can see:

1. Cache the ordering on first computation. Costs space.
2. Change the interface, such that when `with_canonical_atom_ordering` is called, it returns both the canonical clone, as well as the map from old to new id.

I prefer option two, because there is no performance penalty in time or space. Potentially this could be added as a separate method, so we would end up with 1 method for returning just clone, 1 for returning just map of ids, 1 for returning both together. Not sure if it would be worth keeping all three methods indefinitely or just slowly using deprecating 2 of the methods until we just end up with the method that returns both the clone and the map at the same time.

### CODE REVIEW

**Comment by lukasturcani** [comment]
@andrewtarzia

**Comment by andrewtarzia** [comment]
Seen - will do review.

File: src/stk/serialization/json/serializers/constructed_molecule/constructed_molecule.py [hunk 1]
```diff
@@ -107,9 +107,17 @@ def get_keys(building_block):
...
+                return (
+                    None,
+                    None,
+                    None,
+                )
+
             return (
                 building_block_indices[atom_info.get_building_block()],
                 atom_info.get_building_block_id(),
+                atom_info.get_building_block_atom().get_id(),
```
Comment (andrewtarzia) [inline src/stk/serialization/json/serializers/constructed_molecule/constructed_molecule.py:120]: I assume that this ordering does not matter compared to its ordering in the arguments list?

Comment (lukasturcani) [reply src/stk/serialization/json/serializers/constructed_molecule/constructed_molecule.py:120]: Correct, and it means that backwards compatibility of the serialized representation is maintained.

**Review by andrewtarzia — APPROVED** [review]
Only question: is there a test that checks the ordering of a molecule constructed on the fly by the tests?

File: src/stk/serialization/json/serializers/constructed_molecule/constructed_molecule.py (hunk 1 above)
Comment (andrewtarzia) [reply src/stk/serialization/json/serializers/constructed_molecule/constructed_molecule.py:120]: Perfect. Good coding! ahha

**Comment by lukasturcani** [comment]
I think so, if I understand your question the the following tests should cover it?
https://github.com/lukasturcani/stk/pull/267/files#diff-ffe064a35e027491836434ac1a055bf1ec9e262ca5524c682268fe7e4bc99709

**Comment by andrewtarzia** [comment]
Ya I Think they do - just being cautious about the multi-steps being tested, but I think they should be fine!

**Comment by lukasturcani** [comment]
Multi-steps?

**Comment by andrewtarzia** [comment]
I am always concerned when we have an "old" case and a "new" case when the "new" case can undergo an arbitrary number of processes (i.e. reaction, canonicalization....). I think it is fine in this case, but in my head it is hard to picture all of this so I get nervous about whether its covered.

**Comment by lukasturcani** [comment]
I see. Yeah I think that is a good instinct.

**Pull request merged**